src/xdr/Stellar-contract.x

```diff
@@ -0,0 +1,199 @@
+ *
+ *    0x_NNNN_NNNN_NNNN_NNNX  - u63 for any even X
+ *    0x_0000_000N_NNNN_NNN1  - u32
+ *    0x_0000_000N_NNNN_NNN3  - i32
+ *    0x_NNNN_NNNN_NNNN_NNN5  - static: void, true, false, ...
+ *    0x_IIII_IIII_TTTT_TTT7  - object: 32-bit index I, 28-bit type code T
+ *    0x_NNNN_NNNN_NNNN_NNN9  - symbol: up to 10 6-bit identifier characters
+ *    0x_NNNN_NNNN_NNNN_NNNb  - bitset: up to 60 bits
+ *    0x_CCCC_CCCC_TTTT_TTTd  - status: 32-bit code C, 28-bit type code T
+ *    0x_NNNN_NNNN_NNNN_NNNf  - reserved
+ *
+ * Up here in XDR we have variable-length tagged disjoint unions but no
+ * bit-level packing, so we can be more explicit in their structure, at the
+ * cost of spending more than 64 bits to encode many cases, and also having
+ * to convert. It's a little non-obvious at the XDR level why there's a
+ * split between SCVal and SCObject given that they are both immutable types
+ * with value semantics; but the split reflects the split that happens in
+ * the implementation, and marks a place where different implementations of
+ * immutability (CoW, structural sharing, etc.) will likely occur.
+ */
+
+// A symbol is up to 10 chars drawn from [a-zA-Z0-9_], which can be packed
+// into 60 bits with a 6-bit-per-character code, usable as a small key type
+// to specify function, argument, tx-local environment and map entries
+// efficiently.
+typedef string SCSymbol<10>;
+
+enum SCValType
+{
+    SCV_U63 = 0,
+    SCV_U32 = 1,
+    SCV_I32 = 2,
+    SCV_STATIC = 3,
+    SCV_OBJECT = 4,
+    SCV_SYMBOL = 5,
+    SCV_BITSET = 6,
+    SCV_STATUS = 7
+};
+
+% struct SCObject;
+
+enum SCStatic
+{
+    SCS_VOID = 0,
+    SCS_TRUE = 1,
+    SCS_FALSE = 2
+};
+
+enum SCStatusType
+{
+    SST_OK = 0,
+    SST_UNKNOWN_ERROR = 1
+    // TODO: add more
+};
+
+union SCStatus switch (SCStatusType type)
+{
+case SST_OK:
+    void;
+case SST_UNKNOWN_ERROR:
+    uint32 unknownCode;
+};
+
+union SCVal switch (SCValType type)
+{
+case SCV_U63:
+    uint64 u63;
+case SCV_U32:
+    uint32 u32;
+case SCV_I32:
+    int32 i32;
+case SCV_STATIC:
+    SCStatic ic;
+case SCV_OBJECT:
+    SCObject* obj;
+case SCV_SYMBOL:
+    SCSymbol sym;
+case SCV_BITSET:
+    uint64 bits;
+case SCV_STATUS:
+    SCStatus status;
+};
+
+enum SCObjectType
+{
+    // We have a few objects that represent non-stellar-specific concepts
+    // like general-purpose maps, vectors, strings, numbers, blobs, etc.
+
+    SCO_BOX = 0,
+    SCO_VEC = 1,
+    SCO_MAP = 2,
+    SCO_U64 = 3,
+    SCO_I64 = 4,
+    SCO_STRING = 5,
+    SCO_BINARY = 6,
+    SCO_BIGINT = 7,
+    SCO_BIGRAT = 8,
+
+    // Followed by a potentially much longer list of object types
+    // corresponding to any XDR ledger types users might want to manipulate
+    // directly. Separate type codes are required for composite types and
+    // each of their members, if it is reasonable for a user to want to hold
+    // a handle to the member separately from the composite type.
+
+    SCO_LEDGERKEY = 9,
```

Comment:
Why do we need ledger key as a distinct handle type? I would expect there to be an object type for each identifier type we use to key into ledgers. i.e. AccountId, TrustLineLedgerKey, etc. At least on the contract side it seems not particularly useful to have a generic ledger key that could refer to anything.

Reply:
The read-write sets of a contract invocation are sets of LedgerKeys.

Reply:
So is LedgerKeys some subset of OBJs?

Reply:
No, LedgerKeys keys that identify LedgerEntries in the ledger. Read-write sets have to do with concurrency control against the ledger, among multiple concurrent host contexts. There is no concurrent access to host objects.

Reply:
How does a contract use the LedgerKeys object type? When would it encounter the type? What host functions would it use it with?

Reply:
Well, at least my thinking was two contexts, both deriving from the assumption of deterministic concurrency control with ledger keys:

  - When a contract is invoked in a transaction that is bundled together with some set of LedgerKeys as its read-write sets, the contract will then need to _retrieve_ those keys somehow, in order to actually read or write them. It might use a host function to access the sets accompanying the transaction, or it might be passed elements of the sets as contract-function arguments, but either way it seems likely to me that it'd be receiving them as general LedgerKeys since that's how they'd be presented on the transaction's surface, not as specific subtypes of LedgerKey.
  -  Deterministic concurrency control sometimes requires a so-called "reconnaissance query" to run first in order to determine the read-write sets that will be used in a subsequent query with dynamic read-write sets. So in such a reconnaissance query, I was thinking it'd be returning a vector of vectors of LedgerKeys as its result value.
  
I suppose this will depend on whether classic interop winds up having access to the ledger via LedgerKeys at all, or whether it's restricted to CONTRACT_DATA entries subkeyed by SCVal -- in the latter case we'd just need to be indicating CONTRACT_DATA keys for the read-write sets.

Reply:
Hmm okay got it. Is there an assumption that the contract would need to be given a list of read-write sets. Is that necessary? I, maybe naively, thought that the read-write sets would be given to core but not to the contract, and that a contract attempting to read or write outside of the set would be an error. The reason I thought that is I don't imagine writing code in a contract to handle the case where a read-write set missed including something I need. And I imagined that any ledger keys referenced by parameters to the contract call would be simple objects like AccountId.

I think it would be helpful to document some example contract calls which show how this will work.

Reply:
Well, I mean, think it through: suppose your contract is going to read AccountID X. How is it going to _get_ AccountID X into the contract? X isn't going to appear out of thin air. It's going to be provided as a parameter to the contract, right? That is, a transaction that is an invocation of the contract will pass X. But we also want to indicate to core's concurrency-control mechanism (that is going to schedule the execution) that the contract's read-set includes LedgerKey(AccountID(X)). So suppose LedgerKey(AccountID(X)) is the Nth element of the read-set attached to the transaction. We have three-ish options:

  - We invoke the contract passing the Val N, and the contract turns around and calls a host function to fetch its Nth read-set element, in order to acquire LedgerKey(AccountID(X)), and then unwraps AccountID(X) from it.
  - We call the contract with an argument placeholder that says, effectively, "hey contract-invoking path, please grab read-set element N and put it in here as an argument", and then the contract is called with LedgerKey(AccountID(X)) again and can unwrap AccountID(X) from it again.
  - We cal the contract with a Val that holds AccountID(X) directly. Now we've provided 2 copies of AccountID in the transaction: one in the read-set and another redundant one in the contract argument, differing only by "being wrapped in LedgerKey(...)"
 
I figure providing two copies of the same key to a conract is not ideal (indeed if you look in [the project](https://github.com/orgs/stellar/projects/28/views/1) you'll see I have a work item "consolidate invoke tx environment and read-write sets" -- that's what this is about).

I was assuming we'd do the first or second option. But again, maybe all this is irrelevant since we might not be letting contracts access AccountIDs directly at all (they might be restricted to just be using CONTRACT_DATA). This is stuff that I expect will be determined by @jon and others working on the ledger interop plan; I don't feel very strongly about it, I'm just doing stuff here that follows the path I was on on the prototype branch.

Re adding example contracts to this PR: this PR is ahead of any of that. There are no host functions in here at all. It's all work-in-progress to arrive later.

Reply:
If I understand correctly in all three options you describe the read-write set and the arguments are being provided to the contract, however what I was describing was that only the arguments get provided to the contract, and the contract is rather ignorant that a read-write set exists.

For example:

Tx would contain contract invocation with:
- Read-write set elements: `LK(A1)`, `LK(A2)`
- Arguments: `A1`, `A2`, `A3` (A3 is provided here as an argument, but the contract will not actually load its ledger key because it is not referencing a ledger entry)

Contract would receive:
- Arguments: `A1`, `A2`, `A3`

If the contract was provided with A4 without that account being included in the read-write set, and if the contract attempt to load the ledger entry referenced by A4, then the contract invocation would fail.

To highlight: the concern I raised was only with exposing the read-write to the contract. That seems like something that will complicate the contract interface. And I don't think contracts should try to handle cases where their read-write set was insufficient, or use their read-write set as some form of argument. I think that because then people will start using the read-write set list as sort of like another argument list.

To address the issue you raise of duplicate information in the transaction, I think we can address that without exposing ledger keys to contracts. We could say contract invocations provide a list of values, and then the read-write set and arguments are just indexes into that list of values. So the tx would contain a contract invocation like:
- Values: `A1`, `A2`, `A3`, ...
- Read-write set elements: `#1`, `#2`, ... (automatically converted into ledger keys within core)
- Arguments: `#1`, `#2`, `#3`

In this design there is no need to have "LedgerKey" as an object in the host to guest interface, because the guest doesn't concern itself with "ledger entries", it just concerns itself with objects that just happen to be backed by ledger entries.

Reply:
Actually, even with both options 1 and 2 you mentioned I think we can still keep ledger keys out of the contracts themselves by translating a `LedgerKey(AccountID)` in the transaction interface into a `Object(OBJ_ACCOUNT)` in the host to contract interface. I've been assuming we will have `OBJ_ACCOUNT` in addition to `OBJ_ACCOUNTID`.